src/main/scala/com/fulcrumgenomics/util/NcbiRefSeqGffSource.scala

```diff
@@ -0,0 +1,266 @@
+
+  val genomeBuild: Option[String] = {
+    // Make an iterator of generic GFF records that is buffered so we can peek at the next rec
+    val headerLines = _bufferedLines.takeWhile(_.startsWith("#")).toIndexedSeq
+    headerLines.find(_.startsWith("#!genome-build")).map { l => l.split("""\s+""")(1) }
+  }
+
+  {
+    val iter = _bufferedLines.filter(l => !l.startsWith("#")).map { line =>
+      val fs = line.split('\t')
+      GffRecord(fs(0), fs(1), fs(2), int(fs(3)), int(fs(4)), fs(6) == "-", if (fs(7) == ".") -1 else int(fs(7)), fs(8))
+    }.bufferBetter
+
+    while (iter.nonEmpty) {
+      // Scan for the next chromosome record, discard any other records along the way
+      iter.find(rec => rec.kind == "region" && rec.accession.startsWith("NC_")) match {
+        case None =>
+          () // Indicates we've hit the end of the file
+        case Some(rec) =>
+          val chrom = Chrom(rec.accession, rec.end, if (rec.name == "MT") "chrM" else "chr" + rec.name)
+
+          if (this.dict.exists(sd => sd.getSequence(chrom.name) == null)) {
+            iter.dropWhile(_.accession == chrom.accession)
+          }
+          else {
+            chroms += chrom
+
+            // Build a sub-iterator that stops at the end of this chromosome
+            val chromIter = iter.takeWhile(_.accession == chrom.accession)
+
+            // Run through the chromosome's records, parsing out genes
+            chromIter.foreach { rec =>
+              if (rec.kind == "gene") {
+                try {
+                  parseGene(chrom.name, rec, iter).foreach { gene =>
+                    this.genes.get(gene.name) match {
+                      case None =>
+                        this.genes.put(gene.name, gene)
+                      case Some(existing) =>
+                        logger.debug(s"Adding second locus for $gene.")
+                        this.genes.put(gene.name, existing.copy(loci = existing.loci ++ gene.loci))
+                    }
+                  }
+                }
+                catch {
+                  case ex: Exception => logger.error(s"Error parsing gene: $rec"); throw ex
+                }
+              }
+            }
+          }
+      }
+    }
+  }
+  logger.debug(s"Loaded ${genes.size} genes across ${chroms.size} chromosomes.")
+
+  /**
+    * Parses a single gene and all of it's sub-features from the iterator. After invoking this method
+    * the `iter` passed in will be positioned such that the next record is the record _after_ all
+    * children of the gene record (and their children and so on).
+    *
+    * @return a [[Some(Gene)]] if one of more transcripts are parsed, or [[None]] if no transcripts are
+    *         parsed. A [[None]] is primarily returned if the gene has only experimentally determined
+    *         transcript accessions (X?_*), and [[includeXs]] is false.
+    * */
+  private def parseGene(chrom: String, geneRec: GffRecord, iter: BetterBufferedIterator[GffRecord]): Option[Gene] = {
```

Comment:
I don't love all this passing in (and side-effecting) iterators even though they are private methods. Why not just accumulate records and pass around Seqs? It seems like if the records for a region are sequential then it should be easy to break all the records into Seqs and then process them but maybe I'm missing something.

Reply:
I actually like this way of organizing things.  It's a little complicated because of the fact that there's basically hierarchical data represented in a flat structure.  So each function parses a level of the structure.  I say "it's tricky" because there is no easy way to say, e.g. "consume all the records for this gene".  Instead you have to consume the gene records, then for each transcript read the transcript and find all the records that have the transcript as parent, and then repeat for each transcript until you hit the last transcript for the gene.

I'm going to leave it this way for now because it makes sense to me and is working and I'm not convinced the alternative would be clearer.